### PR TITLE
Optimized CRL generation, updating and retrievial

### PR DESCRIPTION
**Pull request by Ceiu**
Several changes to the CRL management, including the removal of the CRLGenerator class, introduction of al's CRL streaming utility and several updates to the CRL frontend classes.

### CODE REVIEW

File: server/src/main/java/org/candlepin/util/X509CRLStreamWriter.java [hunk 1]
```diff
@@ -0,0 +1,785 @@
...
+         */
+        ByteArrayOutputStream temp = new ByteArrayOutputStream();
+
+        int tagNo = DERTags.NULL;
+        Date oldThisUpdate = null;
+        while (true) {
+            int tag = readTag(crlIn, null);
+            tagNo = readTagNumber(crlIn, tag, null);
+
+            if (tagNo == GENERALIZED_TIME || tagNo == UTC_TIME) {
```
Comment (wottop) [inline server/src/main/java/org/candlepin/util/X509CRLStreamWriter.java]: Are we running the risk of getting stuck here? Is there a way for this to miss and not be true?

**Comment by wottop** [comment]
Looks solid to me. Is there a way to keep this from being 35 commits?

**Comment by Ceiu** [comment]
I *suppose* I could squash that down to one commit. :)

File: server/src/main/java/org/candlepin/util/CrlFileUtil.java [hunk 2]
```diff
@@ -14,109 +14,309 @@
...
+            revoke.add(serial.getSerial());
+            serial.setCollected(true);
+        }
+
+        List<BigInteger> unrevoke = new LinkedList<BigInteger>();
+        for (CertificateSerial serial : this.certificateSerialCurator.getExpiredSerials()) {
+            unrevoke.add(serial.getSerial());
+        }
+
+        if (revoke.size() > 0 || revoke.size() > 0) {
```
Comment (cnsnyder) [inline server/src/main/java/org/candlepin/util/CrlFileUtil.java]: Should this be `if (revoke.size() > 0 || unrevoke.size() > 0) {` ?

Comment (Ceiu) [inline server/src/main/java/org/candlepin/util/CrlFileUtil.java]: Yes it should

File: server/src/main/java/org/candlepin/util/DERUtil.java [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+
+    /**
+     * Read the length header from an input stream and return the length read.
+     *
+     * This code is a slightly adapted version of the code in BouncyCastle's
+     * ASN1InputStream.
+     *
+     * @param s the InputStream to read from
+     * @param count the counter to modify.  Can be null.
+     * @return then length of the value
```
Comment (cnsnyder) [inline server/src/main/java/org/candlepin/util/DERUtil.java]: Very minor "then" -> "the" ?

File: server/src/main/java/org/candlepin/util/X509CRLEntryStream.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+ * be to check the signature in hasNext() once revokedSeqBytes hits zero and throw a
+ * RuntimeException.  We would also need a RSA public key as a parameter to the constructor.
+ *
+ * See https://en.wikipedia.org/wiki/X.690 and http://luca.ntop.org/Teaching/Appunti/asn1.html
+ * for reference on ASN1 and DER encoding.
+ */
+public class X509CRLEntryStream implements Closeable, Iterator<X509CRLEntryObject> {
+    private InputStream crlStream;
+
+    // TODO should be a BigInteger?  Apparently long definite lengths can go up to 2^1008 - 1
```
Comment (cnsnyder) [inline server/src/main/java/org/candlepin/util/X509CRLEntryStream.java]: Does this need to be done?

Comment (Ceiu) [inline server/src/main/java/org/candlepin/util/X509CRLEntryStream.java]: @awood

Comment (awood) [inline server/src/main/java/org/candlepin/util/X509CRLEntryStream.java]: > Does this need to be done?

BouncyCastle's ASN1InputStream itself only uses integers for length.  MAXINT for Java is 2,147,483,647 which is roughly 2GB.  Right now the CRL list is at about 80MB so I don't think we'll need a BigInt there.  I can update the comment to indicate that.

File: server/src/main/java/org/candlepin/util/X509CRLStreamWriter.java (hunk 1 above)
Comment (awood) [inline server/src/main/java/org/candlepin/util/X509CRLStreamWriter.java]: > Are we running the risk of getting stuck here? Is there a way for this to miss and not be true?

At some point, we have to hit a GENERALIZED_TIME or UTC_TIME that represents the `thisUpdate` component.  Not hitting one of those would mean the CRL is not compliant with the RFC and corrupt.  Regardless, the code would not run forever because eventually the stream would be emptied by all the read operations and an exception would get thrown.

File: server/src/main/java/org/candlepin/util/DERUtil.java [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+        }
+
+        return tag;
+    }
+
+    public static void writeDER(OutputStream out, byte[] der) throws IOException {
+        writeDER(out, der, null);
+    }
+
+    public static void writeDER(OutputStream out, byte[] der, Signer signer) throws IOException {
```
Comment (cnsnyder) [inline server/src/main/java/org/candlepin/util/DERUtil.java]: This seems to be the same as writeValue(). Why have two methods to do the same thing?

Comment (awood) [inline server/src/main/java/org/candlepin/util/DERUtil.java]: > This seems to be the same as writeValue(). Why have two methods to do the same thing?

They are functionally the same, but semantically different.  `writeValue` is for writing the *V* portion of the TLV and `writeDER` is for writing the entire TLV.  That distinction is not discussed in any comments or otherwise noted.  I'll probably just have `writeDER` delegate to `writeValue` or vice versa to preserve the distinction.  Or perhaps just change them all to `writeBytes`.

**Comment by cnsnyder** [comment]
II've looked it over and it looks good.
I'm presently playing around with it. Below is the only thing I've found so far that seems suspect.

The following is the output of `curl -k -i -u admin:admin --request GET "https://localhost:8443/candlepin/crl"` run just after deploy (before doing anything that would result in an addition to the CRL).
"""
HTTP/1.1 500 Internal Server Error
Server: Apache-Coyote/1.1
x-candlepin-request-uuid: 3db6972e-d256-4b1e-a9d7-7f416474e39c
X-Version: 2.0.10-1
X-Version: 2.0.10-1
Content-Type: application/json
Transfer-Encoding: chunked
Date: Fri, 22 Jan 2016 02:34:08 GMT
Connection: close

{
  "displayMessage" : "/var/lib/candlepin/candlepin-crl.crl (No such file or directory)",
  "requestUuid" : "3db6972e-d256-4b1e-a9d7-7f416474e39c"
}%
"""

**Comment by Ceiu** [comment]
Whoops. Looks like I got cute with optimization and never ensured the file was created. Will be fixed shortly.

**Comment by cnsnyder** [comment]
I'm not sure that the return value should be nothing when the CRL is empty.
Directly after deploy GETting candlepin/crl will return a valid albeit empty X509 CRL (as decoded below) (Please note that the CRL number is 2 and not 1 because this output is from running the command a second time):
"""
Certificate Revocation List (CRL):
        Version 2 (0x1)
    Signature Algorithm: sha1WithRSAEncryption
        Issuer: /CN=8f353fe12557/C=US/L=Raleigh
        Last Update: Jan 22 14:59:28 2016 GMT
        Next Update: Jan 23 14:59:28 2016 GMT
        CRL extensions:
            X509v3 Authority Key Identifier: 
                keyid:D6:1A:29:EF:93:90:CB:F5:3B:A9:3D:5B:42:A4:02:42:43:61:0D:FA
                DirName:/CN=8f353fe12557/C=US/L=Raleigh
                serial:82:13:E3:B7:E1:C3:74:D7

            X509v3 CRL Number: 
                2
No Revoked Certificates.
    Signature Algorithm: sha1WithRSAEncryption
         5a:71:06:7e:90:bd:16:22:b3:f1:5b:40:a0:6d:c5:74:a7:45:
         dd:9f:78:7e:8f:0b:0a:f4:72:19:c2:2f:3f:33:ef:64:da:e9:
         48:c8:5a:44:a5:99:12:d0:4d:de:d5:cb:28:0a:5b:37:47:63:
         22:9b:f7:49:d6:d9:54:8f:39:72:02:b6:83:ef:0a:56:de:bb:
         56:b9:a1:55:fa:f5:56:a1:5f:18:d3:12:df:03:20:bc:95:87:
         8a:7f:31:b0:88:d6:3d:73:dc:ef:c6:30:9c:23:f6:9d:1c:0b:
         9f:fe:14:ce:7c:d0:5e:ef:ba:ca:66:5f:d1:f8:6c:d3:9b:0a:
         b5:e9
-----BEGIN X509 CRL-----
MIIBdjCB4AIBATANBgkqhkiG9w0BAQUFADA2MRUwEwYDVQQDDAw4ZjM1M2ZlMTI1
NTcxCzAJBgNVBAYTAlVTMRAwDgYDVQQHDAdSYWxlaWdoFw0xNjAxMjIxNDU5Mjha
Fw0xNjAxMjMxNDU5MjhaoHYwdDBmBgNVHSMEXzBdgBTWGinvk5DL9TupPVtCpAJC
Q2EN+qE6pDgwNjEVMBMGA1UEAwwMOGYzNTNmZTEyNTU3MQswCQYDVQQGEwJVUzEQ
MA4GA1UEBwwHUmFsZWlnaIIJAIIT47fhw3TXMAoGA1UdFAQDAgECMA0GCSqGSIb3
DQEBBQUAA4GBAFpxBn6QvRYis/FbQKBtxXSnRd2feH6PCwr0chnCLz8z72Ta6UjI
WkSlmRLQTd7VyygKWzdHYyKb90nW2VSPOXICtoPvClbeu1a5oVX69VahXxjTEt8D
ILyVh4p/MbCI1j1z3O/GMJwj9p0cC5/+FM580F7vuspmX9H4bNObCrXp
-----END X509 CRL-----
"""

As it stands now this PR changes this behavior to return nothing. Not sure if that is the appropriate response, thoughts @awood ?

A second difference in behavior from master and this PR is that master increases the CRL number upon each request for the CRL. This PR does not.

**Comment by awood** [comment]
> As it stands now this PR changes this behavior to return nothing. Not sure if that is the appropriate response, thoughts @awood ?

It should probably return a CRL with no entries.  We can easily and cheaply just create one of those on the fly in that situation.

**Comment by cnsnyder** [comment]
@awood  Sounds good to me.
CONDACK given that we return a CRL with no entries when the CRL is requested and has no entries.
(GET /candlpin/crl -> Empty CRL as master returns now)

**Comment by Ceiu** [comment]
```
(11:59:35) crog@cl-vm-1:~/devel/candlepin/server(crog/crl_optimizations) $ sudo rm /var/lib/candlepin/candlepin-crl.crl
(11:59:57) crog@cl-vm-1:~/devel/candlepin/server(crog/crl_optimizations) $ curl -k -u admin:admin "https://localhost:8443/candlepin/crl"
{
  "type" : "X.509",
  "sigAlgName" : "1.2.840.113549.1.1.5",
  "sigAlgParams" : "BQA=",
  "thisUpdate" : "2016-01-25T17:00:12+0000",
  "nextUpdate" : "2016-01-26T17:00:12+0000",
  "revokedCertificates" : null,
  "tbscertList" : "MIHyAgEBMA0GCSqGSIb3DQEBBQUAMD8xHjAcBgNVBAMMFWxvY2FsaG9zdC5sb2NhbGRvbWFpbjELMAkGA1UEBhMCVVMxEDAOBgNVBAcMB1JhbGVpZ2gXDTE2MDEyNTE3MDAxMloXDTE2MDEyNjE3MDAxMlqgfzB9MG8GA1UdIwRoMGaAFI7FRuYDVBrRoxMc5E1L1Xvrsh76oUOkQTA/MR4wHAYDVQQDDBVsb2NhbGhvc3QubG9jYWxkb21haW4xCzAJBgNVBAYTAlVTMRAwDgYDVQQHDAdSYWxlaWdoggkA3nrVYBK7H04wCgYDVR0UBAMCAQA=",
  "issuerX500Principal" : {
    "name" : "L=Raleigh,C=US,CN=localhost.localdomain",
    "encoded" : "MD8xHjAcBgNVBAMMFWxvY2FsaG9zdC5sb2NhbGRvbWFpbjELMAkGA1UEBhMCVVMxEDAOBgNVBAcMB1JhbGVpZ2g="
  },
  "sigAlgOID" : "1.2.840.113549.1.1.5",
  "criticalExtensionOIDs" : [ ],
  "nonCriticalExtensionOIDs" : [ "2.5.29.20", "2.5.29.35" ],
  "signature" : "k3QpUovi9M1q6nR5n2eO1VYt7jQpFyT9Rsw6eYlb/tacFSznLd4BCa/m/DxuCM+N5z7e6d+CNSppDbSq1YTIJwQWXc6YH4EpK3PdpOhAgo9MdK0qdy5mvm994hvipa+CKJoaQC6jZ7vQm4M1hHl/n5+4w5su2sLZthCTG6+d4Uo=",
  "encoded" : "MIIBiDCB8gIBATANBgkqhkiG9w0BAQUFADA/MR4wHAYDVQQDDBVsb2NhbGhvc3QubG9jYWxkb21haW4xCzAJBgNVBAYTAlVTMRAwDgYDVQQHDAdSYWxlaWdoFw0xNjAxMjUxNzAwMTJaFw0xNjAxMjYxNzAwMTJaoH8wfTBvBgNVHSMEaDBmgBSOxUbmA1Qa0aMTHORNS9V767Ie+qFDpEEwPzEeMBwGA1UEAwwVbG9jYWxob3N0LmxvY2FsZG9tYWluMQswCQYDVQQGEwJVUzEQMA4GA1UEBwwHUmFsZWlnaIIJAN561WASux9OMAoGA1UdFAQDAgEAMA0GCSqGSIb3DQEBBQUAA4GBAJN0KVKL4vTNaup0eZ9njtVWLe40KRck/UbMOnmJW/7WnBUs5y3eAQmv5vw8bgjPjec+3unfgjUqaQ20qtWEyCcEFl3OmB+BKStz3aToQIKPTHStKncuZr5vfeIb4qWvgiiaGkAuo2e70JuDNYR5f5+fuMObLtrC2bYQkxuvneFK",
  "issuerDN" : {
    "values" : [ "localhost.localdomain", "US", "Raleigh" ],
    "name" : "CN=localhost.localdomain,C=US,L=Raleigh",
    "encoded" : "MD8xHjAcBgNVBAMMFWxvY2FsaG9zdC5sb2NhbGRvbWFpbjELMAkGA1UEBhMCVVMxEDAOBgNVBAcMB1JhbGVpZ2g=",
    "oids" : [ {
      "id" : "2.5.4.3"
    } ]
  }
curl: (18) NSS: client certificate not found (nickname not specified)
}(12:00:12) crog@cl-vm-1:~/devel/candlepin/server(crog/crl_optimizations) $ 
```

Also added another test to verify we're not getting null, nor an exception, when the file is absent.

**Comment by cnsnyder** [comment]
`curl -k -u admin:admin --request GET "https://192.168.122.1:8443/candlepin/crl"` on a fresh deploy off master returns:
```
-----BEGIN X509 CRL-----
MIIBdjCB4AIBATANBgkqhkiG9w0BAQUFADA2MRUwEwYDVQQDDAxiY2M3YmJiNzc0
MGYxCzAJBgNVBAYTAlVTMRAwDgYDVQQHDAdSYWxlaWdoFw0xNjAxMjUxNzI3MTBa
Fw0xNjAxMjYxNzI3MTBaoHYwdDBmBgNVHSMEXzBdgBTFvLRnIFZvhIuMUueDbCFA
NiyMsqE6pDgwNjEVMBMGA1UEAwwMYmNjN2JiYjc3NDBmMQswCQYDVQQGEwJVUzEQ
MA4GA1UEBwwHUmFsZWlnaIIJAKTNeRYm6yNtMAoGA1UdFAQDAgEBMA0GCSqGSIb3
DQEBBQUAA4GBAGoXwQLhv4WhUWuj4xMjPdTVzqgtNHB9FJQc24MVfU+yH2XwNlIZ
vXfrvOTpf4Z1B3cUW3r/O/lZKiVfBgHYswmPwrv+wd/4SdDx0gG5FOxQKv8FJEXQ
hwVPJdCz5v1DYwfh1RsGXveg5apXDHnri9DQ4UxWXiLV755u7OOKP/nW
-----END X509 CRL-----
```

`curl -k -u admin:admin --request GET "https://192.168.122.1:8443/candlepin/crl"` on a fresh deploy from this branch returns: 
```json
{
  "type" : "X.509",
  "sigAlgName" : "1.2.840.113549.1.1.5",
  "sigAlgParams" : "BQA=",
  "version" : 2,
  "thisUpdate" : "2016-01-25T17:27:10+0000",
  "nextUpdate" : "2016-01-26T17:27:10+0000",
  "revokedCertificates" : null,
  "tbscertList" : "MIHgAgEBMA0GCSqGSIb3DQEBBQUAMDYxFTATBgNVBAMMDDM0MWY5OWE5NTY4ODELMAkGA1UEBhMCVVMxEDAOBgNVBAcMB1JhbGVpZ2gXDTE2MDEyNTE3MjcxMFoXDTE2MDEyNjE3MjcxMFqgdjB0MGYGA1UdIwRfMF2AFH/YnDf0qYUr3Rnx8jekE6B7VYPQoTqkODA2MRUwEwYDVQQDDAwzNDFmOTlhOTU2ODgxCzAJBgNVBAYTAlVTMRAwDgYDVQQHDAdSYWxlaWdoggkAz4d51lYs8xowCgYDVR0UBAMCAQA=",
  "issuerDN" : {
    "values" : [ "341f99a95688", "US", "Raleigh" ],
    "name" : "CN=341f99a95688,C=US,L=Raleigh",
    "encoded" : "MDYxFTATBgNVBAMMDDM0MWY5OWE5NTY4ODELMAkGA1UEBhMCVVMxEDAOBgNVBAcMB1JhbGVpZ2g=",
    "oids" : [ {
      "id" : "2.5.4.3",
      "derencoded" : "BgNVBAM="
    } ]
  }
}
```

This is somewhat peculiar to me as the headers in both include 'Content-Type: application/json' yet master doesn't seem to be returning json.

@Ceiu Will this change break existing consumers of the api?

**Comment by Ceiu** [comment]
Third time's the charm
```
(12:44:33) crog@cl-vm-1:~/devel/candlepin/server(crog/crl_optimizations) $ sudo rm /var/lib/candlepin/candlepin-crl.crl.bak 
(12:44:39) crog@cl-vm-1:~/devel/candlepin/server(crog/crl_optimizations) $ curl -k -u admin:admin "https://localhost:8443/candlepin/crl"
-----BEGIN X509 CRL-----
MIIBiDCB8gIBATANBgkqhkiG9w0BAQUFADA/MR4wHAYDVQQDDBVsb2NhbGhvc3Qu
bG9jYWxkb21haW4xCzAJBgNVBAYTAlVTMRAwDgYDVQQHDAdSYWxlaWdoFw0xNjAx
MjUxNzQ1MjNaFw0xNjAxMjYxNzQ1MjNaoH8wfTBvBgNVHSMEaDBmgBSOxUbmA1Qa
0aMTHORNS9V767Ie+qFDpEEwPzEeMBwGA1UEAwwVbG9jYWxob3N0LmxvY2FsZG9t
YWluMQswCQYDVQQGEwJVUzEQMA4GA1UEBwwHUmFsZWlnaIIJAN561WASux9OMAoG
A1UdFAQDAgEAMA0GCSqGSIb3DQEBBQUAA4GBAG1a9Jp67fm7yAzk3hUAwwqsifYa
gtjrY1viZrVnz7lYcjHan7pU5XB8crdqsxZbYoowxfgyG1YLGJ3KZyrRvi6h7hnO
CAAuT2aG85cwq1eOo/Wcv51/f17WP+V6nUDael1b1+NNzrL1l8TQ1lCR8legsDK6
fBIZRspliP1Gyw/5
-----END X509 CRL-----
(12:45:23) crog@cl-vm-1:~/devel/candlepin/server(crog/crl_optimizations) $ 
```

**Comment by cnsnyder** [comment]
:smile: Looks good! ACK!
